src/ray/gcs/gcs_server/gcs_job_manager.cc

```diff
@@ -114,5 +159,117 @@ void GcsJobManager::HandleReportJobError(const rpc::ReportJobErrorRequest &reque
   GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
 }
 
+void GcsJobManager::HandleSubmitJob(const rpc::SubmitJobRequest &request,
+                                    rpc::SubmitJobReply *reply,
+                                    rpc::SendReplyCallback send_reply_callback) {
+  auto on_done = [reply, send_reply_callback](const Status &status) {
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+  };
+  auto status = SubmitJob(request, on_done);
+  if (!status.ok()) {
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+  }
+}
+
+Status GcsJobManager::SubmitJob(const ray::rpc::SubmitJobRequest &request,
+                                const ray::gcs::StatusCallback &callback) {
+  auto job_id = JobID::FromBinary(request.job_id());
+
+  RAY_LOG(INFO) << "Submitting job " << job_id;
+
+  auto it = jobs_.find(job_id);
+  if (it != jobs_.end()) {
+    RAY_LOG(ERROR) << "Failed to submit job " << job_id << ", job id conflicts.";
+    std::ostringstream ss;
+    ss << "Job id conflicts: " << job_id;
+    return Status::Invalid(ss.str());
+  }
+
+  // Fill in as much information as possible for the dashboard to
+  // show the job info in SUBMITTED state.
```

Comment:
// Set all the fields that we can get from the SubmitJobRequest. So the dashboard can show necessary information even for the jobs that haven't started running yet (still in SUBMITTED state).

Reply:
Fixed.